src/adapters/vertexai/llm.ts

```diff
@@ -0,0 +1,199 @@
+
+  toString(): string {
+    return this.getTextContent();
+  }
+
+  createSnapshot() {
+    return { chunks: shallowCopy(this.chunks) };
+  }
+
+  loadSnapshot(snapshot: ReturnType<typeof this.createSnapshot>): void {
+    Object.assign(this, snapshot);
+  }
+}
+
+export interface VertexAILLMParameters extends GenerateOptions {
+  [key: string]: any;
+  modelParams: GetGenerativeModelParams;
+}
+
+export interface VertexAIGenerateOptions extends GenerateOptions {
+  parameters?: VertexAILLMParameters;
+}
+
+export interface VertexAILLMInput {
+  modelId: string;
+  project: string;
+  location: string;
+  executionOptions?: ExecutionOptions;
+  cache?: LLMCache<VertexAILLMOutput>;
+  parameters?: Record<string, any>;
+}
+
+export class VertexAILLM extends LLM<VertexAILLMOutput> {
+  public readonly emitter: Emitter<GenerateCallbacks> = Emitter.root.child({
+    namespace: ["vertexai", "llm"],
+    creator: this,
+  });
+
+  protected client: GenerativeModel;
+
+  constructor(protected readonly input: VertexAILLMInput) {
+    super(input.modelId, input.executionOptions, input.cache);
+    const vertexAI = new VertexAI({ project: input.project, location: input.location });
+    this.client = vertexAI.getGenerativeModel({
+      model: input.modelId,
+    });
+  }
+
+  static {
+    this.register();
+    Serializer.register(GenerativeModel, {
+      toPlain: (value) => ({
+        project: getPropStrict(value, "project"),
+        location: getPropStrict(value, "location"),
+        model: getPropStrict(value, "model"),
+        safetySettings: getPropStrict(value, "safetySettings"),
+        generationConfig: getPropStrict(value, "generationConfig"),
+      }),
+      fromPlain: (value) => {
+        const vertexAI = new VertexAI({ project: value.project, location: value.location });
+        return vertexAI.getGenerativeModel({
+          model: value.model,
+          safetySettings: value.safetySettings,
+          generationConfig: value.generationConfig,
+        });
+      },
+    });
+  }
+
+  async meta(): Promise<LLMMeta> {
+    // TODO: retrieve data about current model from the given provider API
+    return { tokenLimit: 999 }; // dummy
+  }
+
+  async tokenize(input: LLMInput): Promise<BaseLLMTokenizeOutput> {
+    const response = await this.client.countTokens({
+      contents: [{ parts: [{ text: input }], role: "user" }],
+    });
+    return {
+      tokensCount: response.totalTokens,
+    };
+  }
+
+  protected async _generate(
+    input: LLMInput,
+    options: VertexAIGenerateOptions,
+    run: GetRunContext<this>,
+  ): Promise<VertexAILLMOutput> {
+    const responses = await signalRace(
+      () => this.client.generateContent(input),
+      run.signal,
+      () => {},
+    );
+    let text = "Empty";
+    if (
+      responses.response.candidates != undefined &&
+      responses.response.candidates[0].content.parts[0].text != undefined
+    ) {
+      text = responses.response.candidates[0].content.parts[0].text;
+    }
```

Comment:
- same as previously mentioned
- instead of duplicating the same code, consider creating a `utils.ts` module that would expose functionality that you are reusing over and over again